### ISSUE TITLE
Warning occurs when training VQKD

### ISSUE THREAD

**Issue by Franklin905** [microsoft/unilm#877]
I tried to train a visual tokenizer from scratch with my dataset and the training hyperparameters are as same as that in the paper. 
The number of unused codes was 154 in the first epoch.
However, a warning occurred:
> [W accumulate_grad.h:170] Warning: grad and param do not obey the gradient layout contract. This is not an error, but may impair performance.
> grad.sizes() = [768, 32, 1, 1], strides() = [32, 1, 1, 1]
> param.sizes() = [768, 32, 1, 1], strides() = [32, 1, 32, 32] (function operator())

Then in further epochs, the number of unused codes jumped to 8191 (total codes: 8192) and kept at 8191. It seemed that the model crashed.
I would like to know if this warning results in the crash of the model and how can I eliminate this warning.

By the way, the logger prints the following message every epoch:

> Reset the codebook statistic info in quantizer before each epoch

What does the message mean? Is it right to reset the codebook statistic info before each epoch during training?

**Comment by pengzhiliang** [comment]
Dear:

Thanks for attention~

>The number of unused codes was 154 in the first epoch.

It makes sense because we use K-Means for initialization of codebook.

>[W accumulate_grad.h:170] Warning: grad and param do not obey the gradient layout contract. This is not an error, but may impair performance.
grad.sizes() = [768, 32, 1, 1], strides() = [32, 1, 1, 1]
param.sizes() = [768, 32, 1, 1], strides() = [32, 1, 32, 32] (function operator())

Don't worry. We also encountered it in our before experiments.

> Then in further epochs, the number of unused codes jumped to 8191 (total codes: 8192) and kept at 8191. It seemed that the model crashed.

You are right.

>Reset the codebook statistic info in quantizer before each epoch

We plan to statistic the codebook usage on-fly, so set the usage to 0 at beginning of each epoch. 

As for the crashed issue, I will run the commands we give again for you and post the log a few minutes later.

Hope that can help you.
Zhiliang.

**Comment by pengzhiliang** [comment]
Hello:
  
I run the commands given in the TOKENIZER.md, and get the log:
```
{"train_lr": 9.99640273392222e-06, "train_min_lr": 9.99640273392222e-06, "train_loss": 0.32229475399179996, "train_quant_loss": 0.0014610294104180667, "train_rec_loss": 0.32083372451918873, "train_total_loss": 0.32229475399179996, "train_weight_decay": 9.99999999999955e-05, "train_grad_norm": 0.4644168669954955, "train_Unused_code": 0, "test_loss": 0.28011225389711786, "test_quant_loss": 0.0015745153871125005, "test_rec_loss": 0.27853773878605076, "test_total_loss": 0.28011225389711786, "test_unused_code": 7892, "epoch": 0, "n_parameters": 108829216}
{"train_lr": 2.9997202126383946e-05, "train_min_lr": 2.9997202126383946e-05, "train_loss": 0.2735909906585964, "train_quant_loss": 0.0022095656403733847, "train_rec_loss": 0.27138142498860734, "train_total_loss": 0.2735909906585964, "train_weight_decay": 9.99999999999955e-05, "train_grad_norm": 0.26252581395333907, "train_Unused_code": 7027, "test_loss": 0.26640979956948396, "test_quant_loss": 0.002857453576077453, "test_rec_loss": 0.2635523460692529, "test_total_loss": 0.26640979956948396, "test_unused_code": 7575, "epoch": 1, "n_parameters": 108829216}
{"train_lr": 4.999800151884568e-05, "train_min_lr": 4.999800151884568e-05, "train_loss": 0.26071460531067125, "train_quant_loss": 0.0030147256096778475, "train_rec_loss": 0.25769987966802765, "train_total_loss": 0.26071460531067125, "train_weight_decay": 9.99999999999955e-05, "train_grad_norm": 0.19147715303930732, "train_Unused_code": 6216, "test_loss": 0.2534439312576344, "test_quant_loss": 0.0035439016523234773, "test_rec_loss": 0.249900030141527, "test_total_loss": 0.2534439312576344, "test_unused_code": 7270, "epoch": 2, "n_parameters": 108829216}
{"train_lr": 6.99988009113074e-05, "train_min_lr": 6.99988009113074e-05, "train_loss": 0.2500476030339535, "train_quant_loss": 0.003714332582644387, "train_rec_loss": 0.2463332704654772, "train_total_loss": 0.2500476030339535, "train_weight_decay": 9.99999999999955e-05, "train_grad_norm": 0.1445192570904557, "train_Unused_code": 5697, "test_loss": 0.24337894215502523, "test_quant_loss": 0.004001962111832461, "test_rec_loss": 0.23937698024691959, "test_total_loss": 0.24337894215502523, "test_unused_code": 7114, "epoch": 3, "n_parameters": 108829216}

```
It doesn't tend to crash. 

So you could check your hyper-parameters, because sometimes large lr will lead to crash. And then you should check the teacher model you load is correct. 
According to my experience, VQ-KD is stable for the various teachers.

That is all,
Zhiliang.
